src/main/java/gyro/google/kms/KeyRingResource.java

```diff
@@ -0,0 +1,164 @@
+     */
+    @Required
+    @ValidStrings({
+        "asia-east1",
+        "asia-east2",
+        "asia-northeast1",
+        "asia-northeast2",
+        "asia-northeast3",
+        "asia-south1",
+        "asia-southeast1",
+        "australia-southeast1",
+        "europe-north1",
+        "europe-west1",
+        "europe-west2",
+        "europe-west3",
+        "europe-west4",
+        "europe-west6",
+        "northamerica-northeast1",
+        "us-central1",
+        "us-east1",
+        "us-east4",
+        "us-west1",
+        "us-west2",
+        "us-west3",
+        "southamerica-east1",
+        "eur4",
+        "nam4",
+        "global",
+        "asia",
+        "europe",
+        "us"
+    })
+    public String getLocation() {
+        return location;
+    }
+
+    public void setLocation(String location) {
+        this.location = location;
+    }
+
+    /**
+     * The name of the key ring. Can be letters, numbers, underscores or hyphens. (Required)
+     */
+    @Required
+    @Regex("^([a-z]|[0-9]|-|_)*$")
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    /**
+     * The ID of the key ring.
+     */
+    @Output
+    @Id
+    public String getId() {
+        return id;
+    }
+
+    public void setId(String id) {
+        this.id = id;
+    }
+
+    @Override
+    public void copyFrom(KeyRing model) throws Exception {
+        setId(model.getName());
+        setName(getNameFromId());
+        setLocation(getLocationFromId());
+    }
+
+    @Override
+    protected boolean doRefresh() throws Exception {
+        KeyManagementServiceClient client = createClient(KeyManagementServiceClient.class);
+
+        KeyRing keyRing = client.getKeyRing(getId());
+
+        if (keyRing == null) {
+            return false;
+        }
+
+        copyFrom(keyRing);
+
+        client.shutdownNow();
+
+        return true;
+    }
+
+    @Override
+    protected void doCreate(GyroUI ui, State state) throws Exception {
+        KeyManagementServiceClient client = createClient(KeyManagementServiceClient.class);
+
+        String parent = LocationName.format(getProjectId(), getLocation());
+        KeyRing response = client.createKeyRing(parent, getName(), KeyRing.newBuilder().build());
+
+        copyFrom(response);
+
+        client.shutdownNow();
+    }
+
+    @Override
+    protected void doUpdate(
+        GyroUI ui, State state, Resource current, Set<String> changedFieldNames) throws Exception {
+
+    }
+
+    @Override
+    protected void doDelete(GyroUI ui, State state) throws Exception {
+
+    }
+
+    public String getLocationFromId() {
+        return getId().split("/")[3];
```

Comment:
Do you think splitting based on a string is better than `/` as if they prepend something to the url this will break ?

Reply:
According to the api, the ids of these resources are of the form:
`projects/*/locations/*/keyRings/*` (for the keyRing)
`projects/*/locations/*/keyRings/*/cryptoKeys/` (for the key)
`projects/*/locations/*/keyRings/*/cryptoKeys/*/cryptoKeyVersions/*` (for the version).
I doubt this would be different, but to be on a safer side, we could do something like:
`
List<String> list = Arrays.asList(getId().split("/"));
int index = list.indexOf("locations");
return list.get(index+1);
`
What do you think?

Reply:
That sounds reasonable, see if you can make a utility method as there is a number of places you are splitting the id to get these kind of info